### PR TITLE
Add AOTInductor tutorial

### PR DESCRIPTION
**Pull request by angelayi**
Not necessary for 2.1 release.

### CODE REVIEW

File: prototype_source/aot_inductor_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+#
+# To run the .so, we can import ``aot_inductor_launch`` from
+# ``torch._inductor.utils`` which includes some starter C++ code to launch the
+# program.
+#
+# The runnable program takes in a list of flattened arguments, and returns a
+# list of flattened outputs, so we will need to do a little bit of preprocessing
+# to match the calling/returning convention of the eager module.
+
+from torch._inductor.utils import aot_inductor_launcher
```
Comment (avikchaudhuri) [inline prototype_source/aot_inductor_tutorial.py:116]: I suppose this section could be folded into some utils.

I don't like that you're carrying around `output_spec` from way above where you had to run the model to get it. Is it not available via the `exported_program` itself? If so, why not retrieve it from there?

Then you could basically define a util that takes the `(so_path, exported_program)` output from `aot_compile` and runs it as below, first by loading the container, then running it, then unflattening.

**Review by desertfire — COMMENTED** [review]
Since we are going to let HF try this out, you may need to introduce the special logic for handling kwargs and return type, as you did in `benchmarks/dynamo/common.py`.

File: prototype_source/aot_inductor_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+# .. warning::
+#
+#     AOTInductor and its related features are in prototype status and are
+#     subject to backwards compatibility breaking changes. It is not meant to be
+#     used in any production environment.
+#
+#
+# AOTInductor is a specialized version of
+# `TorchInductor <https://dev-discuss.pytorch.org/t/torchinductor-a-pytorch-native-compiler-with-define-by-run-ir-and-symbolic-shapes/747>`__
+# which take an exported PyTorch model, optimizes it, and generates a shared
```
Comment (svekars) [inline prototype_source/aot_inductor_tutorial.py:19]:
```suggestion
# which takes an exported PyTorch model, optimizes it, and generates a shared
```

File: prototype_source/aot_inductor_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+#
+# AOTInductor is a vital component along the
+# `export <https://pytorch.org/docs/2.1/export.html>`__ path as it provides a way
+# to run an exported model without a Python runtime.
+#
+# In this tutorial, you will learn how take a PyTorch model, export it to a
+# compilation graph, pass it to AOTInductor to be compiled shared binary, and
+# run this binary.
+#
+# Note: a modern NVIDIA GPU (H100, A100, or V100) is recommended for this
```
Comment (svekars) [inline prototype_source/aot_inductor_tutorial.py:31]:
```suggestion
# .. note:: a modern NVIDIA GPU (H100, A100, or V100) is recommended for this
```

File: prototype_source/aot_inductor_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+#   torch._export.aot_compile(
+#       module: Callable,
+#       args: Tuple[Any, ...],
+#       kwargs: Optional[Dict[str, Any]] = None,
+#       *,
+#       dynamic_shapes: Optional[Dict[str, Any]] = None,
+#       options: Optional[Dict[str, Any]] = None,
+#   ) -> Tuple[str, bytes]
+#
+# This function does the following:
```
Comment (svekars) [inline prototype_source/aot_inductor_tutorial.py:55]: Need a blank line to make sure the output is generated correctly.

```suggestion
# This function does the following:
#
```

File: prototype_source/aot_inductor_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+print("Generated C++ Code:")
+with open(f"{so_path[:-3]}.cpp", 'r') as f:
+    print(f.read())
+print()
+
+######################################################################
+# Running the .so
+# ---------------
+#
+# To run the .so, we can import ``aot_inductor_launcher`` from
```
Comment (svekars) [inline prototype_source/aot_inductor_tutorial.py:107]:
```suggestion
# To run the ``.so`` file, we can import ``aot_inductor_launcher`` from
```

**Review by svekars — COMMENTED** [review]
Just a couple of editorial nits. Also, can you please add to the prototype/protoype_index.rst?

File: prototype_source/aot_inductor_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+#       args: Tuple[Any, ...],
+#       kwargs: Optional[Dict[str, Any]] = None,
+#       *,
+#       dynamic_shapes: Optional[Dict[str, Any]] = None,
+#       options: Optional[Dict[str, Any]] = None,
+#   ) -> Tuple[str, bytes]
+#
+# This function does the following:
+#    1. Exports the given model into a compilation graph through the
+#      `torch.export <https://pytorch.org/docs/2.1/export.html>`__ workflow. If
```
Comment (angelayi) [inline prototype_source/aot_inductor_tutorial.py:57]: update to main

File: prototype_source/aot_inductor_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        return self.branch1(x)
+
+def _normalize_inputs(example_inputs) -> Tuple[Tuple[Any], Mapping[str, Any]]:
+    if isinstance(example_inputs, dict):
+        return (), example_inputs
+    else:
+        return example_inputs, {}
+
+with torch.no_grad():
+    m = M().to("cuda")
```
Comment (chauhang) [inline prototype_source/aot_inductor_tutorial.py:87]: Will be good to make the device configurable / passed as a variable to use the example for cpu / mps and other device types

Comment (angelayi) [reply prototype_source/aot_inductor_tutorial.py:87]: I don't think we support the other device types well to include in this tutorial (cc @desertfire)

Comment (desertfire) [reply prototype_source/aot_inductor_tutorial.py:87]: We can run this simple one on cpu, but AOTInductor for cpu is largely untested, and it won't be our focus before PTC.

**Review by desertfire — APPROVED** [review]

File: .jenkins/metadata.json [hunk 8]
```diff
@@ -34,4 +34,7 @@
   "intermediate_source/scaled_dot_product_attention_tutorial.py": {
     "needs": "linux.g5.4xlarge.nvidia.gpu"
   }
```
Comment (svekars) [inline .jenkins/metadata.json:36]:
```suggestion
  },
```

Comment (angelayi) [reply .jenkins/metadata.json]: Ah, sorry about that! Thanks!

File: prototype_source/aot_inductor_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+from torch._inductor.utils import aot_inductor_launcher
+
+aot_inductor_model_container = torch.utils.cpp_extension.load_inline(
+    name="aot_inductor",
+    cpp_sources=[aot_inductor_launcher],
+    functions=["run"],
+    extra_ldflags=[so_path],
+    with_cuda=True,
+)
```
Comment (sayakpaul) [inline prototype_source/aot_inductor_tutorial.py:131]: It's leading:

```bash
    aot_inductor_model_container = load_inline(
  File "/home/sayak/.pyenv/versions/diffusers/lib/python3.10/site-packages/torch/utils/cpp_extension.py", line 1621, in load_inline
    _maybe_write(cpp_source_path, "\n".join(cpp_sources))
TypeError: sequence item 1: expected str instance, function found
```

@angelayi

Comment (sayakpaul) [reply prototype_source/aot_inductor_tutorial.py:131]: I am on torch nightly and CUDA 12.1.

Comment (angelayi) [reply prototype_source/aot_inductor_tutorial.py]: Fixed it! Please let me know if you have any more issues

**Pull request closed without merging**